rp2040-hal/src/i2c/controller/eh1_0_async.rs

```diff
@@ -0,0 +1,365 @@
+use core::{future, marker::PhantomData, ops::Deref, task::Poll};
+use eh_async_1_0_alpha::i2c::{AddressMode, Operation};
+use fugit::HertzU32;
+use rp2040_pac::RESETS;
+
+use crate::{
+    i2c::{Async, Controller, Error, ValidPinScl, ValidPinSda, I2C},
+    pac::i2c0::RegisterBlock,
+    resets::SubsystemReset,
+};
+
+struct PollI2c<'i2c, A, B, F, G, H, T>
+where
+    F: FnMut(&mut I2C<A, B, Controller<Async>>) -> Poll<T>,
+    G: FnMut(&mut I2C<A, B, Controller<Async>>),
+    H: FnMut(&mut I2C<A, B, Controller<Async>>),
+{
+    i2c: &'i2c mut I2C<A, B, Controller<Async>>,
+    is_ready: F,
+    setup_flags: G,
+    cancel: H,
+    running: bool,
+    // captures F's return type.
+    phantom: PhantomData<T>,
+}
+
+impl<'i2c, A, B, F, G, H, T> future::Future for PollI2c<'i2c, A, B, F, G, H, T>
+where
+    F: FnMut(&mut I2C<A, B, Controller<Async>>) -> Poll<T>,
+    G: FnMut(&mut I2C<A, B, Controller<Async>>),
+    H: FnMut(&mut I2C<A, B, Controller<Async>>),
+{
+    type Output = T;
+
+    fn poll(self: core::pin::Pin<&mut Self>, cx: &mut core::task::Context<'_>) -> Poll<T> {
+        // SAFETY: We are not moving anything.
```

Comment:
I don't think this is sufficient to guarantee safety. If I'm not mistaken, creating a `&mut` aliasing some other reference is enough to invoke undefined behavior.

Reply:
Sorry - I didn't properly notice the reason for the `unsafe` block in the first place.
It's not about the reference being unique or not. We already have a unique reference to that data, so we know it's not aliased. It's just that the data is inside a `Pin`.

Reply:
Yes, the concern here is about the pinned/unpinned value.
This is what is done in `poll_fn` so I figured it was safe to do this although I haven't found something that'd explain this.
The doc for pin and unpin aren't very clear to me.